src/bit-systems/link-hover-menu.ts

```diff
@@ -1,47 +1,150 @@
-import { defineQuery, enterQuery, entityExists } from "bitecs";
+import { defineQuery, entityExists } from "bitecs";
 import type { HubsWorld } from "../app";
-import { Interacted, Link, LinkHoverMenu, LinkHoverMenuItem, HoveredRemoteRight } from "../bit-components";
-import { anyEntityWith } from "../utils/bit-utils";
+import { Link, LinkHoverMenu, HoveredRemoteRight, TextTag, Interacted, LinkHoverMenuItem } from "../bit-components";
+import { findAncestorWithComponent, findChildWithComponent } from "../utils/bit-utils";
+import { hubIdFromUrl } from "../utils/media-url-utils";
+import { Text as TroikaText } from "troika-three-text";
+import { handleExitTo2DInterstitial } from "../utils/vr-interstitial";
+import { changeHub } from "../change-hub";
+import { EntityID } from "../utils/networking-types";
+import { setMatrixWorld } from "../utils/three-utils";
+import { LinkType } from "../inflators/link";
 
-const NULL_EID = 0;
-
-const hoveredQuery = defineQuery([HoveredRemoteRight, Link]);
-const hoveredEnterQuery = enterQuery(hoveredQuery);
-const hoveredMenuItemQuery = defineQuery([HoveredRemoteRight, LinkHoverMenuItem]);
+const menuQuery = defineQuery([LinkHoverMenu]);
+const hoveredQuery = defineQuery([HoveredRemoteRight]);
 const clickedMenuItemQuery = defineQuery([Interacted, LinkHoverMenuItem]);
 
-export function linkHoverMenuSystem(world: HubsWorld) {
-  // Assumes always only single LinkHoverMenu entity exists for now.
-  // TODO: Take into account for more than one for VR
-  const menuEid = anyEntityWith(world, LinkHoverMenu)!;
-  const menuObject = world.eid2obj.get(menuEid)!;
-
-  // Save Link object eid in LinkHoverMenu when hovered.
-  hoveredEnterQuery(world).forEach((eid: number) => {
-    LinkHoverMenu.targetObjectRef[menuEid] = eid;
-    const targetObject = world.eid2obj.get(eid)!;
-    targetObject.add(menuObject);
-  });
+function updateLinkMenuTarget(world: HubsWorld, menu: EntityID, sceneIsFrozen: boolean) {
+  if (LinkHoverMenu.targetObjectRef[menu] && !entityExists(world, LinkHoverMenu.targetObjectRef[menu])) {
+    // Clear the invalid entity reference. (The link entity was removed).
+    LinkHoverMenu.targetObjectRef[menu] = 0;
+    return;
+  }
+
+  if (sceneIsFrozen) {
+    LinkHoverMenu.targetObjectRef[menu] = 0;
+    return;
+  }
+
+  const hovered = hoveredQuery(world);
+  const target = hovered.map(eid => findAncestorWithComponent(world, Link, eid))[0] || 0;
```

Comment:
I want to understand the reason why Link and LinkHoverMenu (below) is found from the ancestors because I'm not so familiar with our menu systems. Do we add RemoteHoverTarget component to descendant entities of Link/LinkHoverMenu entities?

Reply:
First we check if we are hovering over the interactable media here and find the link element:
`const target = hovered.find(eid => findAncestorWithComponent(world, Link, eid));`

Then we check if we are hovering over the menu:
`if (hovered.find(eid => findAncestorWithComponent(world, LinkHoverMenu, eid)))`
And as is the Link Button the one catching the focus, we need to look for the parent entity which has the `LinkHoverMenu` element.

So for the first iteration where we look for the Link I think we can just check for `hasComponent`.

Reply:
Sorry I think I wrote my question poorly. What I really wanted to ask was, it seems to be assumed that `Link` entity can be in ancestors, not in descendants, of `LinkHoverMenu` entity, and similarly `LinkHoverMenu` entity can be in ancestors. Who forms or ensures this relation? Similarly,  A certain system? Prefab? Blender add-on?

I don't really like having dependency from an entity to another entity, especially finding a target entity from its ancestors or descendants because it requires new readers to know a lot of additional contexts that are not indicated in a system.

This comment doesn't block this PR. I just wanted to share my thought for future clean up.

Reply:
Looking at this again I don't thing this part makes sense:
```typescript
if (hovered.find(eid => findAncestorWithComponent(world, LinkHoverMenu, eid))) {
    LinkHoverMenu.clearTargetTimer[menu] = world.time.elapsed + 1000;
    return;
  }
```
I wrote this a long time ago so I don't really remember but I think we don't really need to look for the `LinkHoverMenu` in any case so this can be greatly simplified and just get the first hovered link that we find as you have pointed out in another comment. I've updated with that.

Reply:
> I was just testing it and we are not handling that case at all so we probably need to add that in a follow-up PR

Maybe we need preprocessing glTF content